janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphTraversalUtil.java

```diff
@@ -103,45 +109,138 @@ public static JanusGraphTransaction getTx(Traversal.Admin<?, ?> traversal) {
+        return step instanceof MultiQueriable || isMultiQueryCompatibleParent(step);
+    }
+
+    /**
+     * This method closely matches the behavior implemented in <code>LazyBarrierStrategy</code> which ensures that
+     * no <code>NoOpBarrierStep</code>s are inserted if path labels are required. Since the same limitation applies
+     * to <code>JanusGraphMultiQueryStep</code>s, the definition of legal positions for this step is the same.
+     * @param step The step which follows the JanusGraphMultiQueryStep to be inserted.
+     * @return <code>true</code> if no path labels are required for this position, otherwise <code>false</code>.
+     * @see org.apache.tinkerpop.gremlin.process.traversal.strategy.optimization.LazyBarrierStrategy
+     */
+    public static boolean isLegalMultiQueryPosition(Step<?, ?> step) {
+        if (step.getTraversal().getTraverserRequirements().contains(TraverserRequirement.PATH)) return false;
+        boolean labeledPath = false;
+        Step currentStep = step.getTraversal().getStartStep();
+        while (!currentStep.equals(step, true)) {
+            if (step instanceof PathProcessor) {
+                final Set<String> keepLabels = ((PathProcessor) step).getKeepLabels();
+                labeledPath &= keepLabels == null || !keepLabels.isEmpty();
+            }
+            labeledPath |= !step.getLabels().isEmpty();
+            currentStep = currentStep.getNextStep();
+        }
+        return !labeledPath;
+    }
+
+    /**
+     * Starting at the given step, this method searches the traversal backwards to find the most recent step which
+     * matches the given class. Only the traversal of the given step is scanned and parent or child traversals are
+     * not taken into account.
+     * @param stepClass The class of the requested step.
+     * @param start The step from which the search is started.
+     * @param <S> The class of the requested step.
+     * @return An Optional which contains the requested step if it was found.
+     */
+    public static <S> Optional<S> getPreviousStepOfClass(final Class<S> stepClass, Step<?,?> start) {
+        Step currentStep = start;
+        while (currentStep != null && !currentStep.getClass().equals(stepClass) && !(currentStep instanceof EmptyStep)) {
+            currentStep = currentStep.getPreviousStep();
+        }
+        return currentStep != null && currentStep.getClass().equals(stepClass) ? Optional.of((S) currentStep) : Optional.empty();
+    }
+
+    /**
+     * Returns a list of steps from the traversal, which match a given predicate.
+     * @param predicate Whether or not a step should be in the returned list.
+     * @param traversal The traversal whose steps should be used.
+     * @return The list of matching steps.
+     */
+    public static List<Step> getSteps(Predicate<Step> predicate, Traversal.Admin<?,?> traversal) {
+        return traversal.getSteps().stream().filter(predicate).collect(Collectors.toList());
     }
```

Comment:
What is the purpose of collecting Steps into a new List? Will there be concurent modification if we don't collect them?
Otherwise, I would prefer to returning just a stream (i.e. to save cycles from collection overhead):
```
public static Stream<Step> getSteps(Predicate<Step> predicate, Traversal.Admin<?,?> traversal) {
    return traversal.getSteps().stream().filter(predicate);
}
```

Reply:
Yes, the problem is concurrent modification. I initially tried your solution but the collection of steps is used to find the locations where other steps need to be inserted in the traversal.

Reply:
Got it. Thank you for the explanation!